File: swift-bridge/Sources/CoreAnimationBridge/CAReplicatorLayer.swift
import Foundation
import QuartzCore

@_cdecl("ca_replicator_layer_new")
public func ca_replicator_layer_new() -> UnsafeMutableRawPointer? {
    caRetain(CAReplicatorLayer())
}

@_cdecl("ca_replicator_layer_get_instance_count")
public func ca_replicator_layer_get_instance_count(_ handle: UnsafeMutableRawPointer?) -> Int {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return 0 }
    return layer.instanceCount
}

@_cdecl("ca_replicator_layer_set_instance_count")
public func ca_replicator_layer_set_instance_count(_ handle: UnsafeMutableRawPointer?, _ value: Int) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    layer.instanceCount = value
}

@_cdecl("ca_replicator_layer_get_preserves_depth")
public func ca_replicator_layer_get_preserves_depth(_ handle: UnsafeMutableRawPointer?) -> Bool {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return false }
    return layer.preservesDepth
}

@_cdecl("ca_replicator_layer_set_preserves_depth")
public func ca_replicator_layer_set_preserves_depth(_ handle: UnsafeMutableRawPointer?, _ value: Bool) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    layer.preservesDepth = value
}

@_cdecl("ca_replicator_layer_get_instance_delay")
public func ca_replicator_layer_get_instance_delay(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return 0 }
    return layer.instanceDelay
}

@_cdecl("ca_replicator_layer_set_instance_delay")
public func ca_replicator_layer_set_instance_delay(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    layer.instanceDelay = value
}

@_cdecl("ca_replicator_layer_get_instance_transform")
public func ca_replicator_layer_get_instance_transform(_ handle: UnsafeMutableRawPointer?, _ outTransform: UnsafeMutableRawPointer?) -> Bool {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return false }
    return caWriteTransform(layer.instanceTransform, out: outTransform)
}

@_cdecl("ca_replicator_layer_set_instance_transform")
public func ca_replicator_layer_set_instance_transform(_ handle: UnsafeMutableRawPointer?, _ transformRaw: UnsafeMutableRawPointer?) {
    guard let layer: CAReplicatorLayer = caBorrow(handle), let transform = caReadTransform(transformRaw) else { return }
    layer.instanceTransform = transform
}

@_cdecl("ca_replicator_layer_set_instance_color")
public func ca_replicator_layer_set_instance_color(_ handle: UnsafeMutableRawPointer?, _ colorHandle: UnsafeMutableRawPointer?) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    let color: CGColor? = caBorrow(colorHandle)
    layer.instanceColor = color
}

@_cdecl("ca_replicator_layer_get_instance_color")
public func ca_replicator_layer_get_instance_color(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let layer: CAReplicatorLayer = caBorrow(handle), let color = layer.instanceColor else { return nil }
    return caRetain(color)
}

@_cdecl("ca_replicator_layer_get_instance_red_offset")
public func ca_replicator_layer_get_instance_red_offset(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return 0 }
    return layer.instanceRedOffset
}

@_cdecl("ca_replicator_layer_set_instance_red_offset")
public func ca_replicator_layer_set_instance_red_offset(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    layer.instanceRedOffset = value
}

@_cdecl("ca_replicator_layer_get_instance_green_offset")
public func ca_replicator_layer_get_instance_green_offset(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return 0 }
    return layer.instanceGreenOffset
}

@_cdecl("ca_replicator_layer_set_instance_green_offset")
public func ca_replicator_layer_set_instance_green_offset(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    layer.instanceGreenOffset = value
}

@_cdecl("ca_replicator_layer_get_instance_blue_offset")
public func ca_replicator_layer_get_instance_blue_offset(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return 0 }
    return layer.instanceBlueOffset
}

@_cdecl("ca_replicator_layer_set_instance_blue_offset")
public func ca_replicator_layer_set_instance_blue_offset(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    layer.instanceBlueOffset = value
}

@_cdecl("ca_replicator_layer_get_instance_alpha_offset")
public func ca_replicator_layer_get_instance_alpha_offset(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return 0 }
    return layer.instanceAlphaOffset
}

@_cdecl("ca_replicator_layer_set_instance_alpha_offset")
public func ca_replicator_layer_set_instance_alpha_offset(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAReplicatorLayer = caBorrow(handle) else { return }
    layer.instanceAlphaOffset = value
}

File: src/ca_replicator_layer.rs
use std::ops::Deref;

use crate::color::Color;
use crate::layer::{Layer, LayerLike};
use crate::transform::Transform3D;

#[derive(Debug, Clone)]
/// Safe wrapper around `CAReplicatorLayer`. See <https://developer.apple.com/documentation/quartzcore/careplicatorlayer>.
pub struct ReplicatorLayer {
    inner: Layer,
}

impl ReplicatorLayer {
    #[must_use]
    /// Creates a new replicator layer wrapper.
    pub fn new() -> Option<Self> {
        unsafe { Layer::from_raw(crate::ffi::ca_replicator_layer_new()) }
            .map(|inner| Self { inner })
    }

    #[must_use]
    /// Returns the replicator layer's instance count.
    pub fn instance_count(&self) -> usize {
        usize::try_from(unsafe {
            crate::ffi::ca_replicator_layer_get_instance_count(self.as_layer_ptr())
        })
        .unwrap_or_default()
    }

    /// Sets the replicator layer's instance count.
    pub fn set_instance_count(&self, value: usize) {
        let value = isize::try_from(value).unwrap_or(isize::MAX);
        unsafe { crate::ffi::ca_replicator_layer_set_instance_count(self.as_layer_ptr(), value) };
    }

    #[must_use]
    /// Returns whether the replicator layer preserves depth.
    pub fn preserves_depth(&self) -> bool {
        unsafe { crate::ffi::ca_replicator_layer_get_preserves_depth(self.as_layer_ptr()) }
    }

    /// Sets the replicator layer's preserves depth.
    pub fn set_preserves_depth(&self, value: bool) {
        unsafe { crate::ffi::ca_replicator_layer_set_preserves_depth(self.as_layer_ptr(), value) };
    }

    #[must_use]
    /// Returns the replicator layer's instance delay.
    pub fn instance_delay(&self) -> f64 {
        unsafe { crate::ffi::ca_replicator_layer_get_instance_delay(self.as_layer_ptr()) }
    }

    /// Sets the replicator layer's instance delay.
    pub fn set_instance_delay(&self, value: f64) {
        unsafe { crate::ffi::ca_replicator_layer_set_instance_delay(self.as_layer_ptr(), value) };
    }

    #[must_use]
    /// Returns the replicator layer's instance transform.
    pub fn instance_transform(&self) -> Transform3D {
        let mut transform = Transform3D::identity();
        let ok = unsafe {
            crate::ffi::ca_replicator_layer_get_instance_transform(
                self.as_layer_ptr(),
                (&mut transform as *mut Transform3D).cast::<core::ffi::c_void>(),
            )
        };
        if ok {
            transform
        } else {
            Transform3D::identity()
        }
    }

    /// Sets the replicator layer's instance transform.
    pub fn set_instance_transform(&self, transform: Transform3D) {
        unsafe {
            crate::ffi::ca_replicator_layer_set_instance_transform(
                self.as_layer_ptr(),
                (&transform as *const Transform3D).cast::<core::ffi::c_void>(),
            )
        };
    }

    /// Sets the replicator layer's instance color.
    pub fn set_instance_color(&self, color: Option<&Color>) {
        unsafe {
            crate::ffi::ca_replicator_layer_set_instance_color(
                self.as_layer_ptr(),
                color.map_or(core::ptr::null_mut(), Color::as_ptr),
            )
        };
    }

    #[must_use]
    /// Returns the replicator layer's instance color.
    pub fn instance_color(&self) -> Option<Color> {
        unsafe {
            Color::from_raw(crate::ffi::ca_replicator_layer_get_instance_color(
                self.as_layer_ptr(),
            ))
        }
    }

    #[must_use]
    /// Returns the replicator layer's instance red offset.
    pub fn instance_red_offset(&self) -> f32 {
        unsafe { crate::ffi::ca_replicator_layer_get_instance_red_offset(self.as_layer_ptr()) }
    }

    /// Sets the replicator layer's instance red offset.
    pub fn set_instance_red_offset(&self, value: f32) {
        unsafe {
            crate::ffi::ca_replicator_layer_set_instance_red_offset(self.as_layer_ptr(), value)
        };
    }

    #[must_use]
    /// Returns the replicator layer's instance green offset.
    pub fn instance_green_offset(&self) -> f32 {
        unsafe { crate::ffi::ca_replicator_layer_get_instance_green_offset(self.as_layer_ptr()) }
    }

    /// Sets the replicator layer's instance green offset.
    pub fn set_instance_green_offset(&self, value: f32) {
        unsafe {
            crate::ffi::ca_replicator_layer_set_instance_green_offset(self.as_layer_ptr(), value)
        };
    }

    #[must_use]
    /// Returns the replicator layer's instance blue offset.
    pub fn instance_blue_offset(&self) -> f32 {
        unsafe { crate::ffi::ca_replicator_layer_get_instance_blue_offset(self.as_layer_ptr()) }
    }

    /// Sets the replicator layer's instance blue offset.
    pub fn set_instance_blue_offset(&self, value: f32) {
        unsafe {
            crate::ffi::ca_replicator_layer_set_instance_blue_offset(self.as_layer_ptr(), value)
        };
    }

    #[must_use]
    /// Returns the replicator layer's instance alpha offset.
    pub fn instance_alpha_offset(&self) -> f32 {
        unsafe { crate::ffi::ca_replicator_layer_get_instance_alpha_offset(self.as_layer_ptr()) }
    }

    /// Sets the replicator layer's instance alpha offset.
    pub fn set_instance_alpha_offset(&self, value: f32) {
        unsafe {
            crate::ffi::ca_replicator_layer_set_instance_alpha_offset(self.as_layer_ptr(), value)
        };
    }
}

impl Deref for ReplicatorLayer {
    type Target = Layer;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl LayerLike for ReplicatorLayer {
    fn as_layer_ptr(&self) -> *mut core::ffi::c_void {
        self.inner.as_ptr()
    }
}
